src/main/scala/dependencies/model/ArtifactCoordinates.scala

```diff
@@ -18,27 +20,52 @@ abstract class ArtifactCoordinates extends Ordered[ArtifactCoordinates] {
 
 
 object SemanticVersion {
-  private val majorMinorPatch = Pattern.compile("(?<major>[^\\.]+)\\.(?<minor>[^\\.]+)\\.(?<patch>.*)")
+  private val majorMinorPatch = Pattern.compile("(?<major>[^\\.]+)(\\.(?<minor>[^\\.]+)(\\.(?<patch>.*))?)?")
 
   def apply(version: String): SemanticVersion = {
     val matcher = majorMinorPatch.matcher(version)
     if (matcher.matches()) {
-      SemanticVersion(matcher.group("major"), matcher.group("minor"), matcher.group("patch"))
+      SemanticVersion(
+        matcher.group("major"),
+        Strings.nullToEmpty(matcher.group("minor")),
+        Strings.nullToEmpty(matcher.group("patch"))
+      )
     } else {
+      System.err.println(s"Couldn't cleanly parse semantic version: $version")
       SemanticVersion(version, "", "")
     }
   }
+
+  private def comparatorForField(parsedGetter: (SemanticVersion) => Option[java.lang.Long],
+                                 rawGetter: (SemanticVersion) => String): Comparator[SemanticVersion] = {
+    new Comparator[SemanticVersion] {
+      override def compare(o1: SemanticVersion, o2: SemanticVersion): Int =
+        parsedGetter(o1).map(v => v.compareTo(parsedGetter(o2).getOrElse(-1L)))
+            .orElse(Some(rawGetter(o1).compareTo(rawGetter(o2))))
+            .get
+    }
+  }
 }
 
 
 case class SemanticVersion(major: String, minor: String, patch: String)
     extends Ordered[SemanticVersion] {
 
+  private lazy val majorVersionNumber = parse(major)
+  private lazy val minorVersionNumber = parse(minor)
+  private lazy val patchVersionNumber = parse(patch)
+
   override def compare(that: SemanticVersion): Int = ComparisonChain.start()
-      .compare(major, that.major)
-      .compare(minor, that.minor)
-      .compare(patch, that.patch)
+      .compare(this, that, SemanticVersion.comparatorForField(_.majorVersionNumber, _.major))
```

Comment:
I didn't bother to write tests for any of this, but I have a lot of runtime anecdotal data: 

```
$ ./gradlew allDeps --configuration compile | java -jar ~/src/dependencies/target/scala-2.11/dependencies_2.11-1.0-one-jar.jar -o bzl-file
Pinning to latest version 1.12 of MavenArtifact(org.yaml,snakeyaml) from amongst versions: Set(1.12, 1.6)
Pinning to latest version 1.47 of MavenArtifact(com.beust,jcommander) from amongst versions: Set(1.47, 1.27)
Pinning to latest version 2.11.11 of MavenArtifact(org.scala-lang,scala-library) from amongst versions: Set(2.11.7, 2.11.11, 2.11.8)
Pinning to latest version 2.6.0 of MavenArtifact(com.fasterxml.jackson.jaxrs,jackson-jaxrs-json-provider) from amongst versions: Set(2.6.0, 2.3.3)
Pinning to latest version 0.0.57 of MavenArtifact(com.twitter.common,stat-registry) from amongst versions: Set(0.0.57, 0.0.56)
Pinning to latest version 2.2.4 of MavenArtifact(com.google.code.gson,gson) from amongst versions: Set(2.2.4, 2.2.2)
Pinning to latest version 3.1.2 of MavenArtifact(io.dropwizard.metrics,metrics-core) from amongst versions: Set(3.1.1, 3.1.2)
Pinning to latest version 2.9.2 of MavenArtifact(joda-time,joda-time) from amongst versions: Set(1.6.2, 1.6, 2.9.2, 2.2)
Pinning to latest version 1.3.0 of MavenArtifact(net.jpountz.lz4,lz4) from amongst versions: Set(1.3.0, 1.1.0)
Pinning to latest version 4.5 of MavenArtifact(org.apache.httpcomponents,httpclient) from amongst versions: Set(4.2.1, 4.0.1, 4.5, 4.3.6)
Pinning to latest version 0.0.98 of MavenArtifact(com.twitter.common,quantity) from amongst versions: Set(0.0.97, 0.0.98)
Pinning to latest version 4.14.0 of MavenArtifact(com.twitter,scrooge-core_2.11) from amongst versions: Set(4.14.0, 4.7.0)
Pinning to latest version 3.10.5.Final of MavenArtifact(io.netty,netty) from amongst versions: Set(3.10.5.Final, 3.10.1.Final, 3.7.0.Final)
Pinning to latest version 0.0.79 of MavenArtifact(com.twitter.common,jdk-logging) from amongst versions: Set(0.0.79, 0.0.78)
Pinning to latest version 2.6.0 of MavenArtifact(com.fasterxml.jackson.module,jackson-module-jaxb-annotations) from amongst versions: Set(2.3.3, 2.6.0)
Pinning to latest version 2.5.3 of MavenArtifact(com.fasterxml.jackson.module,jackson-module-scala_2.11) from amongst versions: Set(2.5.3, 2.4.4)
Pinning to latest version 1.1.7 of MavenArtifact(ch.qos.logback,logback-classic) from amongst versions: Set(1.1.7, 1.1.2)
Pinning to latest version 2.11.4 of MavenArtifact(org.scala-lang,scala-reflect) from amongst versions: Set(2.11.2, 2.11.4)
Pinning to latest version 0.0.92 of MavenArtifact(com.twitter.common,stats-provider) from amongst versions: Set(0.0.91, 0.0.92)
Pinning to latest version 0.0.72 of MavenArtifact(com.twitter.common,stat) from amongst versions: Set(0.0.72, 0.0.71)
Pinning to latest version 3.0.0 of MavenArtifact(com.google.code.findbugs,jsr305) from amongst versions: Set(2.0.3, 1.3.9, 3.0.0, 2.0.1)
Pinning to latest version 1.2.17 of MavenArtifact(log4j,log4j) from amongst versions: Set(1.2.17, 1.2.16)
Pinning to latest version 3.18.2-GA of MavenArtifact(org.javassist,javassist) from amongst versions: Set(3.18.1-GA, 3.15.0-GA, 3.18.2-GA)
Pinning to latest version 2.6.0 of MavenArtifact(com.fasterxml.jackson.core,jackson-annotations) from amongst versions: Set(2.5.3, 2.6.0, 2.5.0)
Pinning to latest version 19.0 of MavenArtifact(com.google.guava,guava) from amongst versions: Set(19.0, 16.0.1, 18.0)
Pinning to latest version 1.7.5 of MavenArtifact(org.slf4j,slf4j-log4j12) from amongst versions: Set(1.7.5, 1.7.2)
Pinning to latest version 2.6.0 of MavenArtifact(com.fasterxml.jackson.core,jackson-core) from amongst versions: Set(2.5.3, 2.6.0)
Pinning to latest version 0.0.103 of MavenArtifact(com.twitter.common,util-system-mocks) from amongst versions: Set(0.0.103, 0.0.102)
Pinning to latest version 1.9 of MavenArtifact(commons-codec,commons-codec) from amongst versions: Set(1.9, 1.6, 1.5)
Pinning to latest version 1.1.2.6 of MavenArtifact(org.xerial.snappy,snappy-java) from amongst versions: Set(1.1.2.6, 1.0.4.1)
Pinning to latest version 0.0.120 of MavenArtifact(com.twitter.common,util) from amongst versions: Set(0.0.120, 0.0.119)
Pinning to latest version 2.6.0 of MavenArtifact(com.fasterxml.jackson.datatype,jackson-datatype-guava) from amongst versions: Set(2.6.0, 2.5.3)
Pinning to latest version 3.2.2 of MavenArtifact(commons-collections,commons-collections) from amongst versions: Set(3.2.1, 3.2.2)
Pinning to latest version 2.6.0 of MavenArtifact(com.fasterxml.jackson.core,jackson-databind) from amongst versions: Set(2.5.3, 2.6.0)
Pinning to latest version 0.0.89 of MavenArtifact(com.twitter.common,application-action) from amongst versions: Set(0.0.89, 0.0.88)
Pinning to latest version 0.0.109 of MavenArtifact(com.twitter.common,collections) from amongst versions: Set(0.0.109, 0.0.108)
Pinning to latest version 4.4.5 of MavenArtifact(org.apache.httpcomponents,httpcore) from amongst versions: Set(4.2.1, 4.4.5, 4.1.3, 4.4.1, 4.0.1, 4.3.3)
Pinning to latest version 1.2 of MavenArtifact(commons-logging,commons-logging) from amongst versions: Set(1.1.1, 1.2, 1.1.3)
Pinning to latest version 1.1.0.Final of MavenArtifact(javax.validation,validation-api) from amongst versions: Set(1.1.0.Final, 1.0.0.GA)
Pinning to latest version 0.0.66 of MavenArtifact(com.twitter.common,util-executor-service-shutdown) from amongst versions: Set(0.0.65, 0.0.66)
Pinning to latest version 1.2 of MavenArtifact(commons-cli,commons-cli) from amongst versions: Set(1.1, 1.2)
Pinning to latest version 0.0.114 of MavenArtifact(com.twitter.common,base) from amongst versions: Set(0.0.114, 0.0.113)
Pinning to latest version 0.0.114 of MavenArtifact(com.twitter.common,stats) from amongst versions: Set(0.0.114, 0.0.113)
Pinning to latest version 0.9.0 of MavenArtifact(org.apache.thrift,libthrift) from amongst versions: Set(0.5.0-1, 0.9.0, 0.7.0)
Pinning to latest version 0.7.0 of MavenArtifact(net.sourceforge.argparse4j,argparse4j) from amongst versions: Set(0.7.0, 0.4.3)
Pinning to latest version 2.6.0 of MavenArtifact(com.fasterxml.jackson.jaxrs,jackson-jaxrs-base) from amongst versions: Set(2.3.3, 2.6.0)
Pinning to latest version 2.6 of MavenArtifact(jline,jline) from amongst versions: Set(2.6, 0.9.94, 1.0)
Pinning to latest version 2.6 of MavenArtifact(commons-lang,commons-lang) from amongst versions: Set(2.5, 2.4, 2.6)
Pinning to latest version 0.0.77 of MavenArtifact(com.twitter.common,util-sampler) from amongst versions: Set(0.0.77, 0.0.76)
Pinning to latest version 3.5.0-alpha of MavenArtifact(org.apache.zookeeper,zookeeper) from amongst versions: Set(3.4.10, 3.5.0-alpha, 3.4.6)
Pinning to latest version 1.1.7 of MavenArtifact(ch.qos.logback,logback-core) from amongst versions: Set(1.1.7, 1.1.2)
Pinning to latest version 1.7.25 of MavenArtifact(org.slf4j,slf4j-api) from amongst versions: Set(1.7.20, 1.7.7, 1.7.2, 1.7.25, 1.6.1, 1.7.6, 1.7.12, 1.7.5)
Pinning to latest version 2.2.0 of MavenArtifact(com.yammer.metrics,metrics-core) from amongst versions: Set(2.0.3, 2.2.0)
Pinning to latest version 2.5.3 of MavenArtifact(com.fasterxml.jackson.dataformat,jackson-dataformat-yaml) from amongst versions: Set(2.3.3, 2.5.3)
```

Reply:
It's worth mentioning that this is _after_ plucking the "upgraded to" versions from Gradle's `dependencies` output (e.g. "`a:b:1.2.3 -> 1.2.4`")